model/user.go

```diff
@@ -1,8 +1,20 @@
 package model
 
+import (
+	"time"
+
+	"gorm.io/gorm"
+)
+
 type User struct {
-	ID   string
-      Name string `form:"name" validate:"required,excludesall= "`
+	ID        string `gorm:"primaryKey"`
+	CreatedAt time.Time
+	UpdatedAt time.Time
+	DeletedAt gorm.DeletedAt `gorm:"index"`
```

Comment:
```suggestion
	DeletedAt *gorm.DeletedAt `gorm:"index"`
```
DeletedAtはnullの場合もあるので、型をポインタにする必要がある！
もしくは`room.go`で使ってるように`gorm.Model`を使うのもあり！

Reply:
nullの可能性があるものは型をpointerにしないといけないというルールがgormにはあるってことですか？  
また，https://gorm.io/ja_JP/docs/models.html#Embedded-Struct にあるように，gorm.Modelを使うとDeletedAtは型がpointerではないと思うのですがどうですか？

Reply:
`gorm.DeletedAt`がおそらく`*time.Time`になってるはず！
なんでポインタにしなくちゃいけないかというと、go側でnullを扱えるようにするため
例えばint型のカラムがあって値を指定せずにdbにcreateすると、goのintのデフォルト値である0が保存される
これを*int型にすることでNullが保存されるようになる
取り出すときも同様！

Reply:
>なんでポインタにしなくちゃいけないかというと、go側でnullを扱えるようにするため
例えばint型のカラムがあって値を指定せずにdbにcreateすると、goのintのデフォルト値である0が保存される
これを*int型にすることでNullが保存されるようになる
取り出すときも同様！

理解しました．  
調べてみたら，
`DeletedAt`
 ↓
`type DeletedAt sql.NullTime`
↓
```
type NullTime struct {
	Time  time.Time
	Valid bool // Valid is true if Time is not NULL
}
```
こんな感じだったのですが，

Reply:
sqlパッケージはgolangの標準パッケージで、validの部分をfalseにするとデフォルト値ではなくnullになる仕様かな
gorm.DeletedAtに関してはgorm側でvalidの設定を裏でいい感じにやってくれているので、コード書く人はvalidを気にせずtimeの部分だけでnullを扱うことができるようになっているっぽい

ただstringとかintとかに関してはいい感じのものが用意されてないっぽいので、sql.NullStringとかsql.NullInt64とかでvalidもmodelに含めながら頑張るか、ポインタ使うかって感じで、後者が多いかな？
https://qiita.com/taizo/items/3e0b1ca583d8fe2a62a5

Reply:
デフォルト値って言ってるの全部ゼロ値の間違い！

Reply:
ありがとうございます．  
下で指摘しているtokenとかはpoiterを使うように修正しておきます！

Reply:
@ari1021 あとはポインタを使うように修正するだけかな？
https://github.com/ari1021/chat/pull/20/files#r557161853

Reply:
@Khmer495 
これはDeletedAtがgorm側でうまくやってくれているのでpointerにする必要がないという話になっていたのと思うのですがどうでしたか？

Reply:
型がgorm.DeletedなのでポインタじゃなくてOK!